stream-chat-android-client/src/main/java/io/getstream/chat/android/client/setup/state/internal/ClientStateImpl.kt

```diff
@@ -57,6 +58,9 @@ internal class ClientStateImpl(
         _initialized.value = false
         _connectionState.value = ConnectionState.OFFLINE
         _user.value = null
+
+        InitializationCoordinator.getOrCreate()
+            .removeSocketConnectionStateListeners()
```

Comment:
Remove the listeners when we clear state so that there are no stale listeners lying around.

Reply:
Hey @MarinTolic 

Would it be possible to move this out of `ClientStateImpl`?
This is because, from my point of view, `ClientStateImpl.clearState()` is expected to clear the state's data only without calling any side-effect Singleton functions.

Reply:
@kanat 

I guess there are two ways that I see this being possible. One is to manually call it after `ClientState.clearState()` is called, wherever it is called. It's a sort of a roundabout way of doing it and we would have to be mindful of doing it in the future if anything changes, but it would get the job of removing it from the function done. 

The other one is a suggestion from @adasiewiczr to use `WeakReference` for the listeners in order to allow the GC to collect the listeners once individual instnaces of `ChannelStateLogicImpl` are removed from the heap.

I personally prefer going the explicit route of calling remove, but either way you think is the better way of doing this is fine by me.

Of course, if you have any alternative suggestions I'm open to them as well.

Reply:
By `exposting` I meant that `clearState` should be an internal function.

Reply:
We can remove listeners in `ChatClient::disconnect`. 
 `ClientState::clear` should be internal indeed as someone might call it by mistake and will leave our SDK in a weird state. We should change that in a separate PR

Reply:
Switched to removing listeners in `ChatClient::disconnect` as per suggestion.